src/hivpy/circumcision.py

```diff
@@ -0,0 +1,182 @@
+        self.circ_red_20_30 = self.c_data.circ_red_20_30.sample()
+        self.circ_red_30_50 = self.c_data.circ_red_30_50.sample()
+        self.prob_birth_circ = self.c_data.prob_birth_circ.sample()
+
+    def init_birth_circumcision_all(self, population, date):
+        """
+        Initialise circumcision at birth for the entire male population,
+        both born and unborn. COVID disruption is not factored in.
+        """
+        male_population = population.index[population[col.SEX] == SexType.Male]
+        r = rng.uniform(size=len(male_population))
+        circumcision = r < self.prob_birth_circ
+        population.loc[male_population, col.CIRCUMCISED] = circumcision
+        # split newly circumcised population into born and unborn
+        circ_born_population = population.index[population[col.CIRCUMCISED]
+                                                & (population[col.AGE] >= 0.25)]
+        circ_unborn_population = population.index[population[col.CIRCUMCISED]
+                                                  & (population[col.AGE] < 0.25)]
+        # use current simulation start date as circumcision date for born individuals
+        population.loc[circ_born_population, col.CIRCUMCISION_DATE] = date
+        # find date where each unborn individual's age would be 0.25
+        population.loc[circ_unborn_population,
+                       col.CIRCUMCISION_DATE] = population[col.AGE].transform(
+                                                lambda x: date - timedelta(days=(x+0.25)*365))
+
+    def init_birth_circumcision_born(self, population, date):
+        """
+        Initialise circumcision at birth for all born males.
+
+        This is an alternative birth circumcision initialisation
+        method to `init_birth_circumcision_all` and requires
+        the use of `update_birth_circumcision` at every time step
+        to work as expected.
+        """
+        male_born_population = population.index[(population[col.SEX] == SexType.Male)
+                                                & (population[col.AGE] >= 0.25)]
+        r = rng.uniform(size=len(male_born_population))
+        circumcision = r < self.prob_birth_circ
+        population.loc[male_born_population, col.CIRCUMCISED] = circumcision
+        # all circumcised males get a circumcision date of the start of the simulation
+        population.loc[population[col.CIRCUMCISED], col.CIRCUMCISION_DATE] = date
+
+    def update_birth_circumcision(self, population, time_step, date):
+        """
+        Update birth circumcision for newly born males.
+        COVID disruption is factored in.
+
+        Requires birth circumcision to be initialised with
+        `init_birth_circumcision_born` instead of
+        `init_birth_circumcision_all` in order to work as expected.
+        """
+        # covid disruption causes circumcision probability to be 0
+        if (self.covid_disrup_affected != 1) & (self.vmmc_disrup_covid != 1):
```

Comment:
`covid_disrup_affected` and `vmmc_disrup_covid` should probably be booleans since they can only be 1 or 0, and it will also simplify `if` statements